cpp/src/arrow/column.h

```diff
@@ -78,6 +80,9 @@ class ARROW_EXPORT Column {
 
   // @returns: the column's data as a chunked logical array
   std::shared_ptr<ChunkedArray> data() const { return data_; }
+
+  bool Equals(const std::shared_ptr<Column>& other) const;
```

Comment:
Can you make the API for these instead `const ChunkedArray& other` and `const Column& other`? More flexible that way. we should try to do this consistently across the codebase. May make sense to maintain a wrapper function that takes `const shared_ptr<T>&` (sort of like `DataType::Equals`) that checks for nullptr like here